#! /bin/bash

#SBATCH --partition=Orion
#SBATCH --ntasks=4
#SBATCH --cpus-per-task=15
#SBATCH --time=5-00:00:00
#SBATCH --mem=200GB
#SBATCH --mail-user=cnnamdi@charlotte.edu
#SBATCH --mail-type=ALL

echo "======================================================"
echo "Start Time  : $(date)"
echo "Submit Dir  : $SLURM_SUBMIT_DIR"
echo "Job ID/Name : $SLURM_JOBID : $SLURM_JOB_NAME"
echo "Node List   : $SLURM_JOB_NODELIST"
echo "Num Tasks   : $SLURM_NTASKS total [$SLURM_NNODES nodes @ $SLURM_CPUS_ON_NODE CPUs/node]"
echo "======================================================"
echo ""

#=================================================================================
# download and unzip genomes: download interactively in the DTN node
#=================================================================================

: << 'COMMENT'

wget --show-progress -i nonsoftmasked_genomes_download_path.txt
gunzip *fa.gz

#=================================================================================
# softmasking of the genome with RepeatMasker
#=================================================================================

module load repeatmasker/4.1.6
module load parallel/20240422

cd softmasked_genomes

process_file() {
	RepeatMasker -e rmblast \
		-pa 15 \
		-species "sorghum bicolor" \
		-xsmall \
		-libdir /projects/cooper_research2/chinaza/databases \
		$1
}

export -f process_file

file_path=/projects/cooper_research2/chinaza/structural_variant_calling/graph_cactus/genomes/*.fa
find $file_path | parallel -j 4 --joblog job_log.txt process_file {}

#=================================================================================
# select only the 10 chromosomes, exclude scaffolds
#=================================================================================

for file in genomes/*fa.masked
do
	filename=`basename $file .masked`
	echo $file
	echo $filename
	awk '/^>/ {n++} n<=10' $file > cleaned_softmasked_genomes/$filename.masked_cleaned
done

#=================================================================================
# distance estimation between genomes using mash
#=================================================================================

module load anaconda3/2023.09

#conda create -n mash_env
conda activate mash_env

#conda install bioconda::mash
#mash --help

cd cleaned_softmasked_genomes/
#mash sketch -o sketch *.masked_cleaned

#mash info sketch.msh

mash dist -t sketch.msh sketch.msh > distance_matrix.txt

#=================================================================================
# phylogenetic tree construction from distance matrix generated by mash
#=================================================================================

module load anaconda3/2023.09

#conda create -n phylip_env
conda activate phylip_env

#conda install bioconda::phylip

#kitsch

# 'kitsch' was ran interactively because phylip couldn't detect the distance matrix file.
# the file from mash was modified so phylip could read it. This modification include:
#    - replacing the first row with the total number of species (13)
#    - making sure each species name is <= 10 characters long, spaces was 
#	used to make up species name that were less than 10 characters long  
# this new modified file was saved as distance_matrix_modified.txt
# kitsch method was ran with default parameters
# kitsch method was used over fitch and neighbor-joining (NJ) method because cactus does
#	not support polytomies and NJ method resulted in a polytomy. 

#=================================================================================
# multi-genome alignment using progressive cactus
#=================================================================================

cactus_dir="/projects/cooper_research2/chinaza/packages/cactus/cactus-bin-v2.9.3/"
source "${cactus_dir}venv-cactus-v2.9.3/bin/activate"
#cactus --version

#cactus "${cactus_dir}./js" cleaned_softmasked_genomes/seq_file.txt alignment.hal

# important job stats incase of another run: ntasks=4, cpus-per-task=15
# seff 1093407

halStats alignment.hal #slurm-1150629.out

#=================================================================================
# convert hal to vg format using hal2vg
#=================================================================================

cactus_dir="/projects/cooper_research2/chinaza/packages/cactus/cactus-bin-v2.9.3/"
source "${cactus_dir}venv-cactus-v2.9.3/bin/activate"

# hal2vg --version

hal2vg alignment.hal --noAncestors --hdf5InMemory --chop 32 --progress > alignment.vg

# slurm file: slurm-1169111.out, slurm-1188851.out
# task is memory intensive: seff 1169111, 1188851

COMMENT

#=================================================================================
# index generation using toil-vg index: failed (slurm-1412119.out)
#=================================================================================

module load samtools/1.19
module load anaconda3/2023.09

conda activate toilvenv_2

#toil-vg index jobstore output --graphs alignment.vg --chroms 1 2 3 4 5 6 7 8 9 10 --all_index
#slurm-1412119.out

#toil-vg construct -h #slurm-3734594.out

toil-vg index -h #slurm-3734680.out

#=================================================================================
#
#=================================================================================

#=================================================================================
#
#=================================================================================

#=================================================================================
#
#=================================================================================

#=================================================================================
#
#=================================================================================

#=================================================================================
#
#=================================================================================

#=================================================================================
#
#=================================================================================


echo ""
echo "======================================================"
echo "End Time   : $(date)"
echo "======================================================"
